contracts/DAOSignApp.sol

```diff
@@ -0,0 +1,204 @@
+                    posis[data.message.signatureCIDs[i]].message.signer,
+                'Invalid sign proofs'
+            );
+        }
+        return true;
+    }
+
+    function storeProofOfAuthority(SignedProofOfAuthority memory data) external {
+        require(recover(data.message, data.signature) == data.message.from, 'Invalid signature');
+        require(validate(data), 'Invalid message');
+        poaus[data.proofCID].message.name = data.message.name;
+        poaus[data.proofCID].message.from = data.message.from;
+        poaus[data.proofCID].message.agreementCID = data.message.agreementCID;
+        poaus[data.proofCID].message.app = data.message.app;
+        poaus[data.proofCID].message.timestamp = data.message.timestamp;
+        poaus[data.proofCID].message.metadata = data.message.metadata;
+        for (uint i = 0; i < data.message.signers.length; i++) {
+            poauSignersIdx[data.proofCID][data.message.signers[i].addr] = i;
+            poaus[data.proofCID].message.signers.push(data.message.signers[i]);
+        }
+        poaus[data.proofCID].signature = data.signature;
+        poaus[data.proofCID].proofCID = data.proofCID;
+        proof2signer[data.proofCID] = data.message.from;
+        emit NewProofOfAuthority(data);
+    }
+
+    function storeProofOfSignature(SignedProofOfSignature memory data) external {
+        require(recover(data.message, data.signature) == data.message.signer, 'Invalid signature');
+        require(validate(data), 'Invalid message');
+
+        posis[data.proofCID] = data;
+        proof2signer[data.proofCID] = data.message.signer;
+        emit NewProofOfSignature(data);
+    }
+
+    function storeProofOfAgreement(SignedProofOfAgreement memory data) external {
+        require(validate(data), 'Invalid message');
+
+        poags[data.proofCID] = data;
+        emit NewProofOfAgreement(data);
+    }
+
+    function getProofOfAuthority(
+        string memory cid
+    ) external view returns (SignedProofOfAuthorityMsg memory) {
+        SignedProofOfAuthority memory data = poaus[cid];
+        return
+            SignedProofOfAuthorityMsg({
+                message: toEIP712Message(data.message),
+                signature: data.signature
+            });
+    }
+
+    function getProofOfSignature(
+        string memory cid
+    ) external view returns (SignedProofOfSignatureMsg memory) {
+        SignedProofOfSignature memory data = posis[cid];
+        return
+            SignedProofOfSignatureMsg({
+                message: toEIP712Message(data.message),
+                signature: data.signature
+            });
+    }
+
+    function getProofOfAgreement(
+        string memory cid
+    ) external view returns (SignedProofOfAgreementMsg memory) {
+        SignedProofOfAgreement memory data = poags[cid];
+        return
+            SignedProofOfAgreementMsg({
+                message: toEIP712Message(data.message),
+                signature: data.signature
+            });
+    }
+}
```

Comment:
There is no signature of Proof-of-Agreement. It shouldn't be signed by anyone

Reply:
This is a generalized structure, needed for uniformity of answers